src/metadata.rs

```diff
@@ -0,0 +1,261 @@
+//! YubiKey Device Metadata
+
+// Adapted from yubico-piv-tool:
+// <https://github.com/Yubico/yubico-piv-tool/>
+//
+// Copyright (c) 2014-2016 Yubico AB
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+//   * Redistributions of source code must retain the above copyright
+//     notice, this list of conditions and the following disclaimer.
+//
+//   * Redistributions in binary form must reproduce the above
+//     copyright notice, this list of conditions and the following
+//     disclaimer in the documentation and/or other materials provided
+//     with the distribution.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use crate::{consts::*, error::Error, serialization::*, transaction::Transaction, Buffer};
+use std::{ptr, slice};
+use zeroize::Zeroizing;
+
+/// Get metadata item
+pub(crate) fn get_item(data: &[u8], tag: u8) -> Result<&[u8], Error> {
+    let data_len = data.len();
+    let mut p_temp: *const u8 = data.as_ptr();
+    let mut cb_temp: usize = 0;
+    let mut tag_temp: u8;
+
+    unsafe {
+        while p_temp < data.as_ptr().add(data_len) {
+            tag_temp = *p_temp;
+            p_temp = p_temp.add(1);
+            let p_slice =
+                slice::from_raw_parts(p_temp, data.as_ptr() as usize + data_len - p_temp as usize);
+
+            if !has_valid_length(
+                p_slice,
+                data.as_ptr().add(data_len) as usize - p_temp as usize,
+            ) {
+                return Err(Error::SizeError);
+            }
+
+            p_temp = p_temp.add(get_length(p_slice, &mut cb_temp));
+
+            if tag_temp == tag {
+                return Ok(slice::from_raw_parts(p_temp, cb_temp));
```

Comment:
Will take a look

Reply:
Yeah ok, this needs a fix...

Reply:
I made it closer to the original C code in a9d7996, however I don't think it would be possible for that check to ever fail after the `break`?

Really looking forward to rewriting this with safe Rust code eventually.